packages/fx-core/resource/yaml-schema.json

```diff
@@ -0,0 +1,1495 @@
+  "definitions": {
+    "lifeCycleArray": {
+      "type": "array",
+      "items": {
+        "anyOf": [
+          { "$ref": "#/definitions/aadAppCreate" },
+          { "$ref": "#/definitions/aadAppUpdate" },
+          { "$ref": "#/definitions/armDeploy" },
+          { "$ref": "#/definitions/azureStorageEnableStaticWebsite" },
+          { "$ref": "#/definitions/cliRunNpmCommand" },
+          { "$ref": "#/definitions/cliRunDotnetCommand" },
+          { "$ref": "#/definitions/cliRunNpxCommand" },
+          { "$ref": "#/definitions/azureStorageDeploy" },
+          { "$ref": "#/definitions/azureAppServiceDeploy" },
+          { "$ref": "#/definitions/azureFunctionsDeploy" },
+          { "$ref": "#/definitions/teamsAppCreate" },
+          { "$ref": "#/definitions/teamsAppValidateManifest" },
+          { "$ref": "#/definitions/teamsAppValidateAppPackage" },
+          { "$ref": "#/definitions/teamsAppZipAppPackage" },
+          { "$ref": "#/definitions/teamsAppUpdate" },
+          { "$ref": "#/definitions/teamsAppPublishAppPackage" },
+          { "$ref": "#/definitions/botAadAppCreate" },
+          { "$ref": "#/definitions/botframeworkCreate" },
+          { "$ref": "#/definitions/fileCreateOrUpdateEnvironmentFile" },
+          { "$ref": "#/definitions/fileCreateOrUpdateJsonFile" },
+          { "$ref": "#/definitions/fileUpdateEnv" },
+          { "$ref": "#/definitions/fileUpdateJson" },
+          { "$ref": "#/definitions/prerequisiteInstall" },
+          { "$ref": "#/definitions/m365TitleAcquire" },
+          { "$ref": "#/definitions/spfxDeploy" },
+          { "$ref": "#/definitions/teamsAppCopyAppPackageToSPFx" },
+          { "$ref": "#/definitions/script" }
+        ]
+      }
+    },
+    "aadAppCreateBase": {
+      "type": "object",
+      "description": "Create Azure Active Directory application and client secret (optional). Please refer to https://aka.ms/teamsfx-actions/aadapp-create for more details.",
+      "required": ["uses", "writeToEnvironmentFile"],
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "An optional name of this action"
+        },
+        "env": {
+          "type": "object",
+          "description": "Define environment variables",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
+        "uses": {
+          "type": "string",
+          "description": "Create Azure Active Directory application when the environment variable that stores clientId is empty. Also create client secret for the application when generateClientSecret parameter is true and the environment variable that stores clientSecret is empty. When creating new Azure Active Directory application, this action generates clientId, objectId, tenantId, authority and authorityHost. When creating new client secret, this action generates clientSecret. Please refer to https://aka.ms/teamsfx-actions/aadapp-create for more details.",
```

Comment:
Where is this environment variable for `clientId`? Is it TTK environment or shell environment? What if its not empty?

Reply:
This is the `aadAppCreateBase` which is referenced by `aadAppCreateWithSecret` and `aadAppCreateWithoutSecret` as a base "class" to share common fields(see "allOf" in L93).
Schema of clientId is defined in L122 and L182.
If it's not empty, a new AAD app will be created and the new AAD appid will be stored in the envrionment variable specified by clientID. @blackchoey

Reply:
aadAppCreate is of type either `aadAppCreateWithSecret` or `aadAppCreateWithoutSecret`. We implement it this way to provider better intellisense.